iceoryx_dds/source/iceoryx_dds/dds/fastdds_data_reader.cpp

```diff
@@ -0,0 +1,287 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iceoryx_dds/dds/fastdds_data_reader.hpp"
+#include "iceoryx_dds/internal/log/logging.hpp"
+#include "MempoolPubSubTypes.h"
+
+#include <fastdds/dds/domain/DomainParticipantFactory.hpp>
+#include <fastdds/dds/subscriber/qos/DataReaderQos.hpp>
+
+
+iox::dds::FastDDSDataReader::FastDDSDataReader(
+        iox::capro::IdString_t serviceId,
+        iox::capro::IdString_t instanceId,
+        iox::capro::IdString_t eventId) noexcept
+    : m_serviceId(serviceId)
+    , m_instanceId(instanceId)
+    , m_eventId(eventId)
+    , m_type(new Mempool::ChunkPubSubType())
+    , m_listener(nullptr)
+{
+    m_listener.m_fastddsDataReader = this;
+    LogDebug() << m_logName << "Created FastDDSDataReader";
+}
+
+iox::dds::FastDDSDataReader::~FastDDSDataReader()
+{
+    if (m_dataReader != nullptr)
+    {
+        m_subscriber->delete_datareader(m_dataReader);
+    }
+    if (m_subscriber != nullptr)
+    {
+        m_participant->delete_subscriber(m_subscriber);
+    }
+    if (m_topic != nullptr)
+    {
+        m_participant->delete_topic(m_topic);
+    }
+    if (m_participant != nullptr)
+    {
+        eprosima::fastdds::dds::DomainParticipantFactory::get_instance()->delete_participant(m_participant);
+    }
+    LogDebug() << m_logName << "Destroyed FastDDSDataReader";
+}
+
+void iox::dds::FastDDSDataReader::connect() noexcept
+{
+    if (!m_isConnected.load(std::memory_order_relaxed))
+    {
+        // Create the DomainParticipant
+        eprosima::fastdds::dds::DomainParticipantQos participantQos;
+        m_participant =
+            eprosima::fastdds::dds::DomainParticipantFactory::get_instance()->create_participant(0, participantQos);
+
+        // Register the type
+        m_type.register_type(m_participant);
+
+        // Create topic
+        auto topicString =
+            "/" + std::string(m_serviceId) + "/" + std::string(m_instanceId) + "/" + std::string(m_eventId);
+        m_topic = m_participant->create_topic(topicString, "Mempool::Chunk", eprosima::fastdds::dds::TOPIC_QOS_DEFAULT);
+
+        // Create subscriber
+        m_subscriber = m_participant->create_subscriber(eprosima::fastdds::dds::SUBSCRIBER_QOS_DEFAULT, nullptr);
+
+        // Create datareader
+        eprosima::fastdds::dds::DataReaderQos dataReaderQos = eprosima::fastdds::dds::DATAREADER_QOS_DEFAULT;
+        dataReaderQos.history().kind = eprosima::fastdds::dds::KEEP_ALL_HISTORY_QOS;
+        dataReaderQos.endpoint().history_memory_policy =
+            eprosima::fastrtps::rtps::PREALLOCATED_WITH_REALLOC_MEMORY_MODE;
+        m_dataReader = m_subscriber->create_datareader(m_topic, dataReaderQos, &m_listener);
+
+        // Save state
+        m_isConnected.store(true, std::memory_order_relaxed);
```

Comment:
Could it be that another thread is waiting for `m_isConnected` to be true and then accesses the members of this class. If so we have here two problems.
The compiler can reorder the `m_isConnected.store(true)` statement since it has the relaxed memory and all the members are not synchronized which could lead to pink elephants.

If it can happen that one thread calls connect and another thread is then working on that class I would recommend to change the memory order from relaxed to `std::memory_order_seq_cst` for all atomic calls on this atomic - or use the acquire release pattern.